File: src/main/java/com/tkr/desk/BenchmarkTracker.java
package com.tkr.desk;

import com.tkr.types.WireTypes.*;
import com.tkr.types.WireTypes.Status;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Benchmark tracker with active weights and tracking error computation.
 * Ported from legacy-cpp benchmark_tracker.cc.
 */
public final class BenchmarkTracker {

    public static final class BenchmarkConstituent {
        public int symbolId;
        public int weightBp;
        public long benchmarkReturnBp;
    }

    public static final class PortfolioHolding {
        public int accountId;
        public int symbolId;
        public long qtyMilli;
        public long marketValueCents;
    }

    public static final class TrackingErrorEntry {
        public int symbolId;
        public long benchmarkWeightBp;
        public long portfolioWeightBp;
        public long activeWeightBp;
        public long contributionBp;
    }

    public static final class BenchmarkTrackerConfig {
        public int benchmarkId;
        public int rebalanceThresholdBp;

        public BenchmarkTrackerConfig(int benchmarkId, int rebalanceThresholdBp) {
            this.benchmarkId = benchmarkId;
            this.rebalanceThresholdBp = rebalanceThresholdBp;
        }
    }

    public static final class BenchmarkTrackerResult {
        public Status status = Status.OK;
        public List<TrackingErrorEntry> constituents = new ArrayList<>();
        public long portfolioReturnBp;
        public long benchmarkReturnBp;
        public long activeReturnBp;
        public long trackingErrorBp;
        public boolean rebalanceNeeded;
    }

    private final BenchmarkTrackerConfig config;
    private List<BenchmarkConstituent> benchmark = new ArrayList<>();

    public BenchmarkTracker() {
        this(new BenchmarkTrackerConfig(1, 100));
    }

    public BenchmarkTracker(BenchmarkTrackerConfig config) {
        this.config = config != null ? config : new BenchmarkTrackerConfig(1, 100);
    }

    public void setBenchmark(List<BenchmarkConstituent> constituents) {
        benchmark = constituents != null ? new ArrayList<>(constituents) : new ArrayList<>();
    }

    public void clearBenchmark() {
        benchmark.clear();
    }

    public long lookupBenchmarkWeight(int symbolId) {
        for (BenchmarkConstituent c : benchmark) {
            if (c.symbolId == symbolId) return c.weightBp;
        }
        return 0;
    }

    public long lookupBenchmarkReturn(int symbolId) {
        for (BenchmarkConstituent c : benchmark) {
            if (c.symbolId == symbolId) return c.benchmarkReturnBp;
        }
        return 0;
    }

    public static long computePortfolioWeight(PortfolioHolding holding, long totalValueCents) {
        if (totalValueCents == 0) return 0;
        return holding.marketValueCents * 10000L / totalValueCents;
    }

    public static long computeWeightedReturn(List<PortfolioHolding> holdings, List<BenchmarkConstituent> benchmark) {
        long totalValue = 0;
        for (PortfolioHolding h : holdings) totalValue += h.marketValueCents;
        if (totalValue == 0) return 0;
        long weightedReturn = 0;
        for (PortfolioHolding h : holdings) {
            long weight = h.marketValueCents * 10000L / totalValue;
            long symbolReturn = 0;
            for (BenchmarkConstituent c : benchmark) {
                if (c.symbolId == h.symbolId) {
                    symbolReturn = c.benchmarkReturnBp;
                    break;
                }
            }
            weightedReturn += weight * symbolReturn / 10000L;
        }
        return weightedReturn;
    }

    public static long computeTrackingError(List<TrackingErrorEntry> entries) {
        long sumSq = 0;
        for (TrackingErrorEntry e : entries) {
            long diff = e.activeWeightBp;
            sumSq += diff * diff / 10000L;
        }
        long lo = 0;
        long hi = sumSq;
        while (lo < hi) {
            long mid = (lo + hi + 1) / 2;
            if (mid * mid <= sumSq) lo = mid;
            else hi = mid - 1;
        }
        return lo;
    }

    public List<TrackingErrorEntry> buildActiveWeights(List<PortfolioHolding> holdings) {
        List<TrackingErrorEntry> entries = new ArrayList<>();
        long totalValue = 0;
        for (PortfolioHolding h : holdings) totalValue += h.marketValueCents;
        Map<Integer, Long> portfolioWeights = new HashMap<>();
        for (PortfolioHolding h : holdings) {
            portfolioWeights.put(h.symbolId, computePortfolioWeight(h, totalValue));
        }
        for (BenchmarkConstituent c : benchmark) {
            TrackingErrorEntry entry = new TrackingErrorEntry();
            entry.symbolId = c.symbolId;
            entry.benchmarkWeightBp = c.weightBp;
            entry.portfolioWeightBp = portfolioWeights.getOrDefault(c.symbolId, 0L);
            entry.activeWeightBp = entry.portfolioWeightBp - entry.benchmarkWeightBp;
            entry.contributionBp = entry.activeWeightBp * c.benchmarkReturnBp / 10000L;
            entries.add(entry);
        }
        for (PortfolioHolding h : holdings) {
            boolean inBenchmark = false;
            for (BenchmarkConstituent c : benchmark) {
                if (c.symbolId == h.symbolId) {
                    inBenchmark = true;
                    break;
                }
            }
            if (!inBenchmark) {
                TrackingErrorEntry entry = new TrackingErrorEntry();
                entry.symbolId = h.symbolId;
                entry.portfolioWeightBp = computePortfolioWeight(h, totalValue);
                entry.benchmarkWeightBp = 0;
                entry.activeWeightBp = entry.portfolioWeightBp;
                entries.add(entry);
            }
        }
        return entries;
    }

    public BenchmarkTrackerResult track(List<PortfolioHolding> holdings) {
        BenchmarkTrackerResult result = new BenchmarkTrackerResult();
        result.status = Status.OK;
        result.constituents = buildActiveWeights(holdings);
        result.portfolioReturnBp = computeWeightedReturn(holdings, benchmark);
        long benchReturn = 0;
        for (BenchmarkConstituent c : benchmark) {
            benchReturn += (long) c.weightBp * c.benchmarkReturnBp / 10000L;
        }
        result.benchmarkReturnBp = benchReturn;
        result.activeReturnBp = result.portfolioReturnBp - result.benchmarkReturnBp;
        result.trackingErrorBp = computeTrackingError(result.constituents);
        for (TrackingErrorEntry e : result.constituents) {
            if (e.activeWeightBp > config.rebalanceThresholdBp
                    || e.activeWeightBp < -config.rebalanceThresholdBp) {
                result.rebalanceNeeded = true;
                break;
            }
        }
        return result;
    }

    public BenchmarkTrackerResult trackBatch(BatchWireFrame frame) {
        List<PortfolioHolding> holdings = new ArrayList<>();
        for (WireBatchRecord rec : frame.records) {
            PortfolioHolding h = new PortfolioHolding();
            h.accountId = rec.accountId;
            h.symbolId = rec.symbolId;
            h.qtyMilli = rec.qtyMilli;
            h.marketValueCents = (long) rec.qtyMilli * rec.priceTick / 1000L;
            holdings.add(h);
        }
        return track(holdings);
    }

    public void addConstituent(int symbolId, int weightBp, long returnBp) {
        BenchmarkConstituent c = new BenchmarkConstituent();
        c.symbolId = symbolId;
        c.weightBp = weightBp;
        c.benchmarkReturnBp = returnBp;
        benchmark.add(c);
    }

    public List<BenchmarkConstituent> getBenchmark() {
        return new ArrayList<>(benchmark);
    }

    public long maxActiveWeightBp(BenchmarkTrackerResult result) {
        long max = 0;
        for (TrackingErrorEntry e : result.constituents) {
            long abs = Math.abs(e.activeWeightBp);
            if (abs > max) max = abs;
        }
        return max;
    }
}

File: legacy-cpp/src/desk/benchmark_tracker.cc
#include "tkr/desk/benchmark_tracker.h"

#include <algorithm>
#include <unordered_map>

namespace tkr {
namespace desk {

BenchmarkTracker::BenchmarkTracker(BenchmarkTrackerConfig config)
    : config_(config) {}

void BenchmarkTracker::SetBenchmark(
    const std::vector<BenchmarkConstituent>& constituents) {
  benchmark_ = constituents;
}

void BenchmarkTracker::ClearBenchmark() { benchmark_.clear(); }

std::int64_t BenchmarkTracker::LookupBenchmarkWeight(
    std::uint32_t symbol_id) const {
  for (const BenchmarkConstituent& c : benchmark_) {
    if (c.symbol_id == symbol_id) {
      return static_cast<std::int64_t>(c.weight_bp);
    }
  }
  return 0;
}

std::int64_t BenchmarkTracker::LookupBenchmarkReturn(
    std::uint32_t symbol_id) const {
  for (const BenchmarkConstituent& c : benchmark_) {
    if (c.symbol_id == symbol_id) {
      return c.benchmark_return_bp;
    }
  }
  return 0;
}

std::int64_t BenchmarkTracker::ComputePortfolioWeight(
    const PortfolioHolding& holding, std::int64_t total_value_cents) const {
  if (total_value_cents == 0) {
    return 0;
  }
  return (holding.market_value_cents * 10000) / total_value_cents;
}

std::int64_t BenchmarkTracker::ComputeWeightedReturn(
    const std::vector<PortfolioHolding>& holdings,
    const std::vector<BenchmarkConstituent>& benchmark) {
  std::int64_t total_value = 0;
  for (const PortfolioHolding& h : holdings) {
    total_value += h.market_value_cents;
  }

  if (total_value == 0) {
    return 0;
  }

  std::int64_t weighted_return = 0;
  for (const PortfolioHolding& h : holdings) {
    const std::int64_t weight = (h.market_value_cents * 10000) / total_value;
    std::int64_t symbol_return = 0;
    for (const BenchmarkConstituent& c : benchmark) {
      if (c.symbol_id == h.symbol_id) {
        symbol_return = c.benchmark_return_bp;
        break;
      }
    }
    weighted_return += (weight * symbol_return) / 10000;
  }

  return weighted_return;
}

std::int64_t BenchmarkTracker::ComputeTrackingError(
    const std::vector<TrackingErrorEntry>& entries) {
  std::int64_t sum_sq = 0;
  for (const TrackingErrorEntry& e : entries) {
    const std::int64_t diff = e.active_weight_bp;
    sum_sq += (diff * diff) / 10000;
  }

  std::int64_t te = 0;
  std::int64_t lo = 0;
  std::int64_t hi = sum_sq;
  while (lo < hi) {
    const std::int64_t mid = (lo + hi + 1) / 2;
    if (mid * mid <= sum_sq) {
      lo = mid;
    } else {
      hi = mid - 1;
    }
  }
  te = lo;
  return te;
}

std::vector<TrackingErrorEntry> BenchmarkTracker::BuildActiveWeights(
    const std::vector<PortfolioHolding>& holdings) const {
  std::vector<TrackingErrorEntry> entries;

  std::int64_t total_value = 0;
  for (const PortfolioHolding& h : holdings) {
    total_value += h.market_value_cents;
  }

  std::unordered_map<std::uint32_t, std::int64_t> portfolio_weights;
  for (const PortfolioHolding& h : holdings) {
    portfolio_weights[h.symbol_id] =
        ComputePortfolioWeight(h, total_value);
  }

  for (const BenchmarkConstituent& c : benchmark_) {
    TrackingErrorEntry entry{};
    entry.symbol_id = c.symbol_id;
    entry.benchmark_weight_bp = static_cast<std::int64_t>(c.weight_bp);
    entry.portfolio_weight_bp = portfolio_weights[c.symbol_id];
    entry.active_weight_bp =
        entry.portfolio_weight_bp - entry.benchmark_weight_bp;
    entry.contribution_bp =
        (entry.active_weight_bp * c.benchmark_return_bp) / 10000;
    entries.push_back(entry);
  }

  for (const PortfolioHolding& h : holdings) {
    bool in_benchmark = false;
    for (const BenchmarkConstituent& c : benchmark_) {
      if (c.symbol_id == h.symbol_id) {
        in_benchmark = true;
        break;
      }
    }
    if (!in_benchmark) {
      TrackingErrorEntry entry{};
      entry.symbol_id = h.symbol_id;
      entry.portfolio_weight_bp =
          ComputePortfolioWeight(h, total_value);
      entry.benchmark_weight_bp = 0;
      entry.active_weight_bp = entry.portfolio_weight_bp;
      entries.push_back(entry);
    }
  }

  return entries;
}

BenchmarkTrackerResult BenchmarkTracker::Track(
    const std::vector<PortfolioHolding>& holdings) {
  BenchmarkTrackerResult result{};
  result.status = Status::kOk;

  result.constituents = BuildActiveWeights(holdings);
  result.portfolio_return_bp =
      ComputeWeightedReturn(holdings, benchmark_);

  std::int64_t bench_return = 0;
  for (const BenchmarkConstituent& c : benchmark_) {
    bench_return += (static_cast<std::int64_t>(c.weight_bp) *
                     c.benchmark_return_bp) /
                    10000;
  }
  result.benchmark_return_bp = bench_return;
  result.active_return_bp =
      result.portfolio_return_bp - result.benchmark_return_bp;
  result.tracking_error_bp =
      ComputeTrackingError(result.constituents);

  for (const TrackingErrorEntry& e : result.constituents) {
    if (e.active_weight_bp > config_.rebalance_threshold_bp ||
        e.active_weight_bp < -static_cast<std::int64_t>(
                                 config_.rebalance_threshold_bp)) {
      result.rebalance_needed = true;
      break;
    }
  }

  return result;
}

BenchmarkTrackerResult BenchmarkTracker::TrackBatch(
    const BatchWireFrame& frame) {
  std::vector<PortfolioHolding> holdings;
  holdings.reserve(frame.records.size());

  for (const WireBatchRecord& rec : frame.records) {
    PortfolioHolding h{};
    h.account_id = rec.account_id;
    h.symbol_id = rec.symbol_id;
    h.qty_milli = static_cast<std::int64_t>(rec.qty_milli);
    h.market_value_cents =
        (static_cast<std::int64_t>(rec.qty_milli) *
         static_cast<std::int64_t>(rec.price_tick)) /
        1000;
    holdings.push_back(h);
  }

  return Track(holdings);
}

}  // namespace desk
}  // namespace tkr
